profiling: add Task/GC/Inference/Codegen signposts for Tracy

topolarity opened the pull request:
This adds a variety of metadata to send to Tracy during profiling, in particular method instance signatures, the Heap size, GC Mark/Sweep/Stop-the-World, and a couple of other small pieces of relevant metadata.

It also adds full support for Tasks, tracking these as "Fibers" on the Tracy side:

Check out an example trace here: https://topolarity.github.io/tracy-traces/tracy-experimental/

![image](https://user-images.githubusercontent.com/84105208/230207523-dfd257fd-ce0c-49db-bdf7-b4c31516615e.png)

##### Loose ends

Before this is ready for final review, I still need to:
 - [x] Finalize the JL_TIMING API changes
 - [x] Resolve incompatibility between `TRACY_ON_DEMAND` and Tracy's Fiber support

topolarity committed:
Add `JL_TIMING` support for Tracy

topolarity committed:
ios: Add fixed-size streams and expand inline storage

This includes a few separate enhancements:
  1. `growable` is added to allow for fixed-size buffers
  2. `ios_t` struct layout is optimized for better packing
  3. inline buffer is increased to 83 bytes from 54

This change grows the size of ios_t by 16 bytes (160 -> 172 bytes), but
the internal buffer was increased by 29 bytes thanks to re-arranging the
fields in the struct to be packed more efficiently.

topolarity committed:
profiling: Add Task/GC/Inference/Codegen signposts for Tracy

This adds a variety of metadata to send to Tracy during profiling, in
particular method instance signatures, the Heap size, and a couple of
other small pieces of relevant metadata.

It also adds full support for Task execution, tracking these as "Fibers"
on the Tracy side. This does incur some overhead until an enhancement
is implemented on the Tracy-side to avoid a globally synchronous Fiber
event queue, but most of the events that we're tracking now are
relatively coarse so it should not be a major issue.

topolarity committed:
Add `WAIT_FOR_TRACY` environment variable

This environmental variable instructs the Julia runtime to wait until
it receives a connection from the Tracy profiler before starting.

This allows us to get a complete trace, without having to accumulate a
large queue of events to send over the wire later.

topolarity committed:
deps: Add `TracyPlotConfig` patch for Tracy

This functionality is currently only exposed on the master branch of
Tracy, but it's quite useful so let's carry the patch to expose it until
the next Tracy release.

topolarity committed:
rtutils: Add "quiet" type-printing for `jl_static_show`

This is useful for Tracy, which can receive method signature strings
over the write. For performance reasons, we want to keep those strings
as small as possible, typically less than 80 characters long.

This change adds basic "configuration" support for `jl_static_show`,
including only a "quiet" parameter for now which can be used to avoid
printing DataType parameters.

topolarity committed:
profiling: Add `WITH_TIMING_COUNTS` build option

This replaces the `ENABLE_TIMINGS` #define in `options.h`

Since we now support multiple back-ends for timing events, each back-end
is expected to be enabled with a `WITH_*` Makefile flag, and then the
JL_TIMING API is automatically enabled if any back-end is available.

Turning on multiple back-ends at the same time is also supported.

topolarity committed:
profiling: limit reported method instances

Cap the number of method instances we report to the profiler for a
single LLVM_MODULE_FINISH to 10.

topolarity committed:
deps: Upgrade LibTracyClient to 0.9.1

gbaraldi commented on src/support/ios.h:
```diff
@@ -19,13 +19,13 @@ extern "C" {
 typedef enum { bst_none, bst_rd, bst_wr } bufstate_t;
 
-#define IOS_INLSIZE 54
+#define IOS_INLSIZE 83
```
Why was this changed?

kpamnany commented on src/jitlayers.cpp:
```diff
@@ -248,9 +248,12 @@ static jl_callptr_t _jl_compile_codeinst(
     for (auto &def : emitted) {
         jl_code_instance_t *this_code = def.first;
-        jl_timing_show_func_sig(this_code->def->specTypes, JL_TIMING_CURRENT_BLOCK);
+        if (i < 10)
```
Maybe make this controllable with an optional environment variable?

topolarity replied:
Any recommendations about what to call it?

How about something fairly coarse like `JULIA_VERBOSE_TIMING_METADATA=1` that would turn off limits like this in general?

staticfloat replied:
JULIA_TIMING_METADATA_PRINT_LIMIT, which defaults to 10?

kpamnany commented on src/jlapi.c:
```diff
@@ -680,6 +684,11 @@ static void rr_detach_teleport(void) {
 {
+#ifdef USE_TRACY
+    if (getenv("WAIT_FOR_TRACY"))
+        while (!TracyCIsConnected) ; // Wait for connection
```
Better a `jl_cpu_pause()` rather than just spinning.

kpamnany commented on src/julia.h:
```diff
@@ -1934,6 +1934,10 @@ typedef struct _jl_task_t {
 
 // hidden state:
+
+#ifdef USE_TRACY
+    const char *name;
+#endif
```
Mentioned on the call, adding this note so it isn't lost: would be great to have a `name` irrespective of Tracy.

kpamnany reviewed (approved):
Looks fantastic!

Another useful point to instrument might be thread sleep/wakeup; search for `PROBE` in `partr.c`.

staticfloat commented on src/jlapi.c:
```diff
@@ -680,6 +684,11 @@ static void rr_detach_teleport(void) {
 
 JL_DLLEXPORT int jl_repl_entrypoint(int argc, char *argv[])
 {
+#ifdef USE_TRACY
```
I think it would be very useful to have an environment variable that allows changing `jl_timing_enable_mask` before we're running Julia user code.  Something like `JULIA_TIMING_SUBSYSTEMS=INFERENCE,GC,METHOD_MATCH,-SYSIMG_LOAD` which then gets decoded into turning on the first three elements and turning the last one off (relative to the default).  That way, we can have sensible defaults for `jl_timing_enable_mask` without limiting flexibility.

topolarity replied:
A great idea! Let me see what I can whip up.

topolarity committed:
deps: bump libtracyclient to v0.9.1+1

This enables broadcast messages so that Tracy processes can be listed
by the profiler, which is a nice quality-of-life feature.

topolarity committed:
timing: add envvars for subsystem enable and verbose metadata

This commit adds the following environment variables:
  - `JULIA_TIMING_SUBSYSTEMS` which can be set to, e.g.,
    "+INFERENCE,-GC,METHOD_MATCH" to enable INFERENCE
    and METHOD_MATCH and disable GC.
  - `JULIA_TIMING_METADATA_PRINT_LIMIT` which defaults
    to 10 and determines how many metadata items to add to
    a single timing zone before truncating

This commit also includes other miscellaneous changes to incorporate
review feedback.

topolarity commented:
> Another useful point to instrument might be thread sleep/wakeup; search for PROBE in partr.c.

Great idea - I've added that to my list for the follow-up PR (which is targeting locks, etc.)

One other detail that I noticed today is that `JULIA_WAIT_FOR_TRACY` interacts poorly with precompilation. The spawned Julia processes will see the environment variable, also wait for a Tracy connection on start-up, and will make no progress.

Any ideas about how we should handle that?

staticfloat commented:
I would just delete that environment variable by adding a mapping to `nothing` [here](https://github.com/JuliaLang/julia/blob/ff7b8eb00bf887f20bf57fb7e53be0070a242c07/base/loading.jl#L2104).

topolarity committed:
profiling: delete `JULIA_WAIT_FOR_TRACY` when pre-compiling

We'll eventually want to provide some kind of useful interface to turn
this on for a particular pre-compilation, but for now we should just
prevent causing pre-compilation to hang forever.

carstenbauer commented:
Great stuff! However, I noticed that this PR doesn't add any documentation for the new features. I think that would be a very valuable addition.
